{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#there are two scripts iro_iuc.py and blaCARB.py which take kleborate's Carbapenemase genes and IUC/IRO/RMP genes and search them against\n",
    "#all assemblies. The results, iuciro.tsv and and carbGenes.tsv contain all the hits.\n",
    "#They are cross referenced against the results of the plasmidfinder run in PlasmidFinderHits which are one per assembly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "from os import listdir\n",
    "from os.path import isfile, join, splitext\n",
    "import numpy as np\n",
    "from collections import Counter\n",
    "import pandas as pd\n",
    "import umap\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio import SeqIO\n",
    "import pickle as pkl\n",
    "\n",
    "inputDir=\"/mnt/storage5/anton/KpReplicons/inputData/\"\n",
    "outputDir=\"/mnt/storage5/anton/KpReplicons/outputData/\"\n",
    "plasmidHits=\"/mnt/storage5/anton/KpReplicons/PlasmidFinderHits/\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load metadata on contigs: length, carbapenemase, sideriphores, assembly, plasmid replicons\n",
    "\n",
    "with open(\"/mnt/storage5/anton/KpReplicons/ContigData.pkl\", 'rb') as pkl_file:\n",
    "    contigsData = pkl.load(pkl_file)\n",
    "\n",
    "files = [f for f in listdir(plasmidHits) if isfile(join(plasmidHits, f))  and (splitext(f)[-1]==\".fasta\" or splitext(f)[-1]==\".fna\") and f[0:3]==\"GCF\"  ]\n",
    "\n",
    "# Reading data from multiple files without known indices into dataframe is slow\n",
    "# instead read it into dictionary and populate dataframe afterwards\n",
    "contigPlasmids={} # key=contig ID, value=[plamid IDs]\n",
    "contigAssembly={} # key=contig ID, value=assembly\n",
    "for file in files:\n",
    "    with open(plasmidHits+file) as plamidFile:\n",
    "        for line in plamidFile:\n",
    "            if line[0]==\">\":\n",
    "                values=line.split(\":\")\n",
    "                values[4]=values[4].split(\" \")[0]\n",
    "                if values[4] not in contigPlasmids:\n",
    "                    contigPlasmids[values[4]]=set()\n",
    "                contigPlasmids[values[4]].add(values[0].replace(\">\",\"\"))\n",
    "                contigAssembly[values[4]]=file\n",
    "\n",
    "contigsMetaData=pd.DataFrame(index=contigsData[\"contigToassembly\"].keys(), columns=[\"Assembly\"])\n",
    "contigsMetaData.index.name=\"Contig\"\n",
    "contigsMetaData[\"Assembly\"]=[contigsData[\"contigToassembly\"][f] for f in contigsMetaData.index]\n",
    "contigsMetaData[\"Length\"]=[contigsData[\"contigLenghts\"][f] for f in contigsMetaData.index]\n",
    "contigsMetaData[\"Replicons\"]=[contigPlasmids[f] if f in contigPlasmids else \"\" for f in contigsMetaData.index]\n",
    "contigsMetaData[\"iro\"]=False\n",
    "contigsMetaData[\"iuc\"]=False\n",
    "contigsMetaData[\"Carbapenemase\"]=\"\"\n",
    "contigsMetaData[\"CarbapenemaseLength\"]=0\n",
    "contigsMetaData[\"HasCarbapenemase\"]=False\n",
    "\n",
    "# iuciro and carbGenes.tsv are both generated using BLAST of assemblies versus Kleborate AMR/siderophore database\n",
    "with open(f'{inputDir}iuciro.tsv') as hvFile:\n",
    "    for line in hvFile:\n",
    "        values=line.split(\"\\t\")        \n",
    "        for siderophore in [\"iuc\",\"iro\"]:\n",
    "            if  line.find(siderophore)>-1 and values[0] in contigsMetaData.index:\n",
    "                contigsMetaData.at[values[0], siderophore]=True\n",
    "\n",
    "with open(f'{inputDir}carbGenes.tsv') as carbFile:\n",
    "    for line in carbFile:\n",
    "        values=line.split(\"\\t\")\n",
    "        if float(values[6])==100 and values[0] in contigsMetaData.index:\n",
    "            if int(values[2])-int(values[1])>contigsMetaData.at[values[0], \"CarbapenemaseLength\"]:\n",
    "                contigsMetaData.at[values[0], \"Carbapenemase\"]=values[3] if contigsMetaData.at[values[0],\"Carbapenemase\"]==\"\" else contigsMetaData.at[values[0], \"Carbapenemase\"]+\";\"+values[3]\n",
    "                contigsMetaData.at[values[0], \"CarbapenemaseLength\"]=int(values[2])-int(values[1])\n",
    "        contigsMetaData.at[values[0], \"HasCarbapenemase\"]=True # see if there are some imperfect matches for carbapenemases\n",
    "\n",
    "# # Drop the contigs without carbapenemases, siderophores or replicons as they are not relevant for subsequent analysis\n",
    "# contigsMetaData=contigsMetaData.loc[ (contigsMetaData[\"iuc\"]) | (contigsMetaData[\"iro\"]) | (contigsMetaData[\"Carbapenemase\"]!=\"\") ]\n",
    "\n",
    "# Add the cluster assignment data to the contigsMetaData\n",
    "embeddingClustersDF=pd.read_csv(f'{outputDir}/embedding.tsv', sep=\"\\t\", index_col=0, header=0)\n",
    "for column in [\"Cluster\",\"Country\",\"ST\",\"CollectionYear\", \"Species\"]:\n",
    "    contigsMetaData[column]=[embeddingClustersDF.at[f,column] if f in embeddingClustersDF.index else \"\"  for f in contigsMetaData[\"Assembly\"]]\n",
    "\n",
    "contigsMetaData[\"Cluster\"].replace(np.nan,\"\",inplace=True)\n",
    "contigsMetaData[\"isChr\"]=contigsMetaData[\"Length\"]>4500000\n",
    "\n",
    "# Drop assemblies that do not have cluster data as they are not relevant for subsequent analysis\n",
    "contigsMetaData=contigsMetaData.loc[ contigsMetaData[\"Cluster\"]!=\"\" ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "clustersDescription=pd.DataFrame(index=set(contigsMetaData[\"Cluster\"]), \n",
    "    columns=[\"N\", \"Main STs\", \"Total STs\", \"Species\",\"Collection Dates\",\"Main Replicons\",\"Main Countries\",\"Siderophores\",\"Carbapenemases\",\n",
    "    \"Contig with CR and replicon\",\"Contig with HV and replicon\", \"Contig with CR and HV\",\n",
    "    \"Chr with CR and replicon\",\"Chr with HV and replicon\", \"Chr with CR and HV\"])\n",
    "\n",
    "for cluster in set(contigsMetaData[\"Cluster\"]):\n",
    "    clustersDescription.at[cluster,\"N\"]=len( set( contigsMetaData.loc[ contigsMetaData[\"Cluster\"]==cluster, \"Assembly\" ] ) )\n",
    "    clustersDescription.at[cluster,\"Total STs\"]=len( set( contigsMetaData.loc[ contigsMetaData[\"Cluster\"]==cluster, \"ST\" ] ) )\n",
    "    clustersDescription.at[cluster,\"Species\"]=', '.join(set( contigsMetaData.loc[ contigsMetaData[\"Cluster\"]==cluster, \"Species\" ] ) )\n",
    "    collectionYears=set(contigsMetaData.loc[ (contigsMetaData[\"CollectionYear\"].notnull()) & (contigsMetaData[\"Cluster\"]==cluster)  , \"CollectionYear\" ])\n",
    "    clustersDescription.at[cluster,\"Collection Dates\"]=str(int(min( collectionYears )))+\"-\"+str(int(max( collectionYears )))\n",
    "    \n",
    "    replicons=[k for f in contigsMetaData.loc[contigsMetaData[\"Cluster\"]==cluster, \"Replicons\"] for k in f]\n",
    "    sortedReplicons={k: v for k, v in sorted(Counter(replicons).items(), key=lambda item: item[1], reverse=True)}\n",
    "    clustersDescription.at[cluster,\"Main Replicons\"]=\", \".join( [key+\"(\"+str(sortedReplicons[key])+\")\" for key in sortedReplicons if sortedReplicons[key]>5 ]  )\n",
    "\n",
    "    #These two columns require a particular style of aggregation\n",
    "    for target, source  in zip([\"Main STs\",\"Main Countries\"],[\"ST\",\"Country\"]):\n",
    "        stsWithDuplicates=contigsMetaData.loc[ (contigsMetaData[source].notnull()) & (contigsMetaData[\"Cluster\"]==cluster)]\n",
    "        assemblySTs={}\n",
    "        for index in stsWithDuplicates.index:\n",
    "            assemblySTs[stsWithDuplicates.at[index,\"Assembly\"]]=stsWithDuplicates.at[index,source]\n",
    "\n",
    "        sortedSTs={k: v for k, v in sorted(Counter(assemblySTs.values()).items(), key=lambda item: item[1], reverse=True)}\n",
    "        clustersDescription.at[cluster,target]=\", \".join( [key+\"(\"+str(sortedSTs[key])+\")\" for key in sortedSTs if sortedSTs[key]>5 ]  )\n",
    "    \n",
    "    # \"Contig with CR and replicon\",\"Contig with HV and replicon\", \"Contig with CR and HV\"\n",
    "    for length, prefix in zip([0,4500000],[\"Contig\",\"Chr\"]):\n",
    "        clustersDescription.at[cluster,f'{prefix} with CR and replicon']=Counter( (contigsMetaData[\"Cluster\"]==cluster) & (contigsMetaData[\"Length\"]>length) &\n",
    "                ( (~contigsMetaData[\"iuc\"]) & (~contigsMetaData[\"iro\"]) ) & \n",
    "                ( contigsMetaData[\"Carbapenemase\"]!=\"\")  & \n",
    "                ([len(f)>0 for f in contigsMetaData[\"Replicons\"]]) )[True]\n",
    "        clustersDescription.at[cluster,f'{prefix} with HV and replicon']=Counter( (contigsMetaData[\"Cluster\"]==cluster) & (contigsMetaData[\"Length\"]>length) &\n",
    "                ( (contigsMetaData[\"iuc\"]) | (contigsMetaData[\"iro\"]) ) & \n",
    "                ( contigsMetaData[\"Carbapenemase\"]==\"\")  & \n",
    "                ([len(f)>0 for f in contigsMetaData[\"Replicons\"]]) )[True]\n",
    "        clustersDescription.at[cluster,f'{prefix} with CR and HV']=Counter( (contigsMetaData[\"Cluster\"]==cluster) & (contigsMetaData[\"Length\"]>length) &\n",
    "                ( (contigsMetaData[\"iuc\"]) | (contigsMetaData[\"iro\"]) ) & \n",
    "                ( contigsMetaData[\"Carbapenemase\"]!=\"\")  & \n",
    "                ([len(f)>0 for f in contigsMetaData[\"Replicons\"]]) )[True]\n",
    "\n",
    "    # Add the siderophores and Bla_Carb_acquired basedon embeddingClustersDF (i.e. kleborate output)\n",
    "    aerobactinAlleles=[f'{key}({str(value)})' for key,value in Counter(embeddingClustersDF.loc[embeddingClustersDF[\"Cluster\"]==cluster][\"Aerobactin\"]).items() if value>5 and key!=\"-\"]\n",
    "    salcmochelinAlleles=[f'{key}({str(value)})' for key,value in Counter(embeddingClustersDF.loc[embeddingClustersDF[\"Cluster\"]==cluster][\"Salmochelin\"]).items() if value>5 and key!=\"-\"]\n",
    "    clustersDescription.at[cluster,\"Siderophores\"]=\",\".join(aerobactinAlleles+salcmochelinAlleles)\n",
    "    carbapenemaseAlleles=[f'{key}({str(value)})' for key,value in Counter(embeddingClustersDF.loc[embeddingClustersDF[\"Cluster\"]==cluster][\"Bla_Carb_acquired\"]).items() if value>5 and key!=\"-\"]\n",
    "    clustersDescription.at[cluster,\"Carbapenemases\"]=\",\".join(carbapenemaseAlleles)\n",
    "\n",
    "\n",
    "\n",
    "clustersDescription.to_csv(f'{outputDir}/clustersDescription.tsv',sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "geneLocations=contigsMetaData.loc[ (contigsMetaData[\"iuc\"]) | (contigsMetaData[\"iro\"] ) | \n",
    "                ( contigsMetaData[\"Carbapenemase\"]!=\"\")  | \n",
    "                ([len(f)>0 for f in contigsMetaData[\"Replicons\"]])]\n",
    "geneLocations.index.name=\"Contig\"\n",
    "geneLocations.to_csv(f'{outputDir}geneLocations.tsv',sep=\"\\t\")"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "a612c00234cb2afc6511c78aaf3593a4092443bb2a0bed4620e32910b743a5d1"
  },
  "kernelspec": {
   "display_name": "Python 3.7.6 ('base': conda)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
